bin/poclcc.c

```diff
@@ -0,0 +1,368 @@
+{
+  char *buffer;
+  size_t size, read_size;
+  FILE *kern_file;
+
+  kern_file = fopen(filename, "r");
+  if (kern_file == NULL)
+    return NULL;
+
+  fseek(kern_file, 0, SEEK_END);
+  size = ftell(kern_file);
+  rewind(kern_file);
+
+  buffer = malloc(size + 1);
+  read_size = fread(buffer, 1, size, kern_file);
+  if (read_size != size)
+    {
+      free(buffer);
+      fclose(kern_file);
+      return NULL;
+    }
+  fclose(kern_file);
+  buffer[size] = '\0';
+
+  return buffer;
+}
+
+int process_kernel_file(int arg, char **argv, int argc)
+{
+  if (arg >= argc)
+    return poclcc_error("Incomplete argument for input file!\n");
+
+  char *filename = argv[arg];
+  char *ext = ".pocl";
+  kernel_source = kernel_load_file(filename);
+  if (output_file == NULL)
+    {
+      output_file = malloc(strlen(filename)+strlen(ext));
+      strcpy(output_file, filename);
+      strcat(output_file, ext);
+    }
+  return 0;
+}
+
+/**********************************************************
+ * OPTIONS PROCESS FUNCTIONS*/
+
+int process_help(int arg, char **argv, int argc)
+{
+  print_help();
+  return 0;
+}
+
+int process_output(int arg, char **argv, int argc)
+{
+  if (arg >= argc)
+    return poclcc_error("Incomplete argument for output file!\n");
+
+  output_file = argv[arg];
+  return 0;
+}
+
+int process_opencl_device(int arg, char **argv, int argc)
+{
+  if (arg >= argc)
+    return poclcc_error("Incomplete argument for device_type!\n");
+
+  char *opencl_string = argv[arg];
+  if (!strcmp(opencl_string, "CL_DEVICE_TYPE_CPU"))
+    opencl_device = CL_DEVICE_TYPE_CPU;
+  else if (!strcmp(opencl_string, "CL_DEVICE_TYPE_GPU"))
+    opencl_device = CL_DEVICE_TYPE_GPU;
+  else if (!strcmp(opencl_string, "CL_DEVICE_TYPE_ACCELERATOR"))
+    opencl_device = CL_DEVICE_TYPE_ACCELERATOR;
+  else if (!strcmp(opencl_string, "CL_DEVICE_TYPE_DEFAULT"))
+    opencl_device = CL_DEVICE_TYPE_DEFAULT;
+  else if (!strcmp(opencl_string, "CL_DEVICE_TYPE_ALL"))
+    opencl_device = CL_DEVICE_TYPE_ALL;
+  else
+    {
+      printf("Invalid argument for device_type!\n");
+      return print_help();
+    }
+  return 0;
+}
+
+int process_build_options(int arg, char **argv, int argc)
+{
+  if (arg >= argc)
+    return poclcc_error("Incomplete argument for build_options!\n");
+
+  build_options = argv[arg];
+  return 0;
+}
+
+int process_device_id(int arg, char **argv, int argc)
+{
+  if (arg >= argc)
+    return poclcc_error("Incomplete argument for build_options!\n");
+
+  opencl_device_id = atoi(argv[arg]);
+  return 0;
+}
+
+int process_list_devices(int arg, char **argv, int argc)
+{
+  list_devices = 1;
+  return 0;
+}
+
+/**********************************************************/
+
+poclcc_option options[NUM_OPTIONS]=
+{
+  {process_help, "-h",
+   "\t-h\n"
+   "\t\tDisplay the help\n",
+   1},
+  {process_build_options, "-b",
+   "\t-b <options>\n"
+   "\t\tBuild the program with <options> options\n",
+   2},
+  {process_opencl_device, "-d",
```

Comment:
Is this needed? If we just have a list of devices (with indexes printed) and then a selector based on the index with a default of selecting the only found device, if any.